File: english_learning/tests/catalog_hardening.test.js
// Phase 1A.5 — generator hardening tests (modes + merge-safety).
//   node tests/catalog_hardening.test.js
"use strict";
const fs = require("fs");
const os = require("os");
const path = require("path");
const cp = require("child_process");
const assert = require("assert");

const ROOT = path.join(__dirname, "..");
const GEN = path.join(ROOT, "tools", "gen_territory_catalog.js");
const gen = require(GEN);   // exported pure fns

function run(mode, dir) {
  return cp.execFileSync(process.execPath, [GEN, mode],
    { env: Object.assign({}, process.env, { TERR_CATALOG_DIR: dir }), encoding: "utf8" });
}
function readT(dir, map) { return JSON.parse(fs.readFileSync(path.join(dir, "territories", map + ".json"), "utf8")); }
function writeT(dir, map, arr) { fs.writeFileSync(path.join(dir, "territories", map + ".json"), JSON.stringify(arr, null, 2)); }
function hashDir(dir) {
  const parts = [];
  function walk(d) { fs.readdirSync(d).sort().forEach(f => { const p = path.join(d, f); const s = fs.statSync(p); if (s.isDirectory()) walk(p); else parts.push(f + ":" + fs.readFileSync(p, "utf8")); }); }
  walk(dir); return parts.join("\n");
}

let n = 0; function test(name, fn) { fn(); n++; console.log("  ok -", name); }

// ---- pure backfill: designer fields survive, missing fields added, _meta preserved ----
test("backfill preserves designer fields and back-fills missing ones", () => {
  const disc = { id: "china:pBJ", mapId: "china", regionCode: "pBJ", displayName: "GEN NAME",
    localizedNames: { en: "GenName" }, svgPathKeys: ["pBJ"], administrativeCode: null,
    gamePopulation: 999, metadata: {} };
  const existing = { id: "china:pBJ", mapId: "china", regionCode: "pBJ",
    displayName: "Designer Name", localizedNames: { en: "Designer" }, svgPathKeys: ["pBJ"],
    gamePopulation: 123, terrainType: "mountain",              // designer edits
    _meta: { schemaVersion: 1, catalogVersion: 1, generatedBy: "x", generatedAt: "T0",
             generatorVersion: "0.9", lastDesignerEdit: "2026-01-01" } };
  gen.backfill(existing, disc, "NOW");
  assert.strictEqual(existing.displayName, "Designer Name", "displayName not overwritten");
  assert.strictEqual(existing.gamePopulation, 123, "gamePopulation not overwritten");
  assert.strictEqual(existing.terrainType, "mountain", "terrain preserved");
  assert.strictEqual(existing.economy, null, "missing placeholder back-filled");
  assert.deepStrictEqual(existing.quests, [], "missing placeholder back-filled");
  assert.strictEqual(existing._meta.generatedAt, "T0", "generatedAt preserved");
  assert.strictEqual(existing._meta.lastDesignerEdit, "2026-01-01", "lastDesignerEdit preserved");
  assert.strictEqual(existing._meta.generatorVersion, gen.GENERATOR_VERSION, "generatorVersion refreshed");
});

// ---- CLI modes on a temp catalog dir ----
const tmp = fs.mkdtempSync(path.join(os.tmpdir(), "wd-"));
fs.mkdirSync(path.join(tmp, "territories"), { recursive: true });

test("init creates missing files with _meta + placeholders", () => {
  run("--init", tmp);
  assert.ok(fs.existsSync(path.join(tmp, "catalog.json")));
  const china = readT(tmp, "china");
  assert.ok(china.length > 0);
  const t = china[0];
  assert.ok(t._meta && t._meta.schemaVersion === gen.SCHEMA_VERSION, "_meta present");
  assert.ok("economy" in t && "quests" in t && "ai" in t && "events" in t, "placeholders present");
});

test("init does NOT overwrite an existing file (designer edit survives)", () => {
  const china = readT(tmp, "china");
  china[0].gamePopulation = 4242; china[0].terrainType = "desert";
  china[0]._meta.lastDesignerEdit = "2026-02-02";
  writeT(tmp, "china", china);
  run("--init", tmp);                                     // should skip (file exists)
  const after = readT(tmp, "china");
  assert.strictEqual(after[0].gamePopulation, 4242, "init left designer value untouched");
  assert.strictEqual(after[0].terrainType, "desert");
  assert.strictEqual(after[0]._meta.lastDesignerEdit, "2026-02-02");
});

test("merge preserves designer edits and re-adds a removed placeholder", () => {
  const china = readT(tmp, "china");
  delete china[0].economy;                                // simulate schema drift / missing field
  writeT(tmp, "china", china);
  run("--merge", tmp);
  const after = readT(tmp, "china");
  assert.strictEqual(after[0].gamePopulation, 4242, "merge preserved designer population");
  assert.strictEqual(after[0].terrainType, "desert", "merge preserved terrain");
  assert.strictEqual(after[0]._meta.lastDesignerEdit, "2026-02-02", "merge preserved lastDesignerEdit");
  assert.strictEqual(after[0].economy, null, "merge re-added missing placeholder");
});

test("check is read-only (no file modifications)", () => {
  const before = hashDir(tmp);
  try { run("--check", tmp); } catch (e) { /* non-zero exit on drift is allowed; we only assert no writes */ }
  const after = hashDir(tmp);
  assert.strictEqual(before, after, "check must not modify any file");
});

console.log("\nAll " + n + " hardening tests passed.");

File: english_learning/tests/catalog_hardening_test.py
#!/usr/bin/env python3
"""Phase 1A.5 — validator hardening tests (schema/metadata ownership).

    python3 tests/catalog_hardening_test.py
"""
import json
import os
import sys
import tempfile

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, ROOT)
from territory_catalog import TerritoryCatalog  # noqa: E402

passed = 0


def ok(name):
    global passed
    passed += 1
    print("  ok -", name)


def meta(sv=1):
    return {"schemaVersion": sv, "catalogVersion": 1, "generatedBy": "gen",
            "generatedAt": "T0", "generatorVersion": "1.0.0", "lastDesignerEdit": None}


def terr(tid, mid, code, keys=None, sv=1, with_meta=True):
    t = {"id": tid, "mapId": mid, "regionCode": code, "displayName": code,
         "localizedNames": {"en": code}, "svgPathKeys": keys or [code],
         "administrativeCode": None, "gamePopulation": 100,
         "populationSource": {"value": None, "year": None, "sourceName": None, "sourceUrl": None},
         "metadata": {}, "terrainType": None, "settlementType": None, "features": [],
         "adjacentTerritoryIds": [], "economy": None, "quests": [], "ai": None, "events": []}
    if with_meta:
        t["_meta"] = meta(sv)
    return t


def build(maps, terrs, catalog=None):
    d = tempfile.mkdtemp()
    os.makedirs(os.path.join(d, "territories"), exist_ok=True)
    json.dump(maps, open(os.path.join(d, "maps.json"), "w"))
    if catalog is not None:
        json.dump(catalog, open(os.path.join(d, "catalog.json"), "w"))
    for mid, ts in terrs.items():
        json.dump(ts, open(os.path.join(d, "territories", mid + ".json"), "w"))
    return TerritoryCatalog(base_dir=d).load()


CAT = {"catalogVersion": 1, "schemaVersion": 1, "generatorVersion": "1.0.0", "maps": 1, "territories": 1}
MAPS = [{"id": "m", "svgFile": "maps/m.svg", "name": "M"}]

# 1. real catalog still validates clean
real = TerritoryCatalog().load()
assert real.validate() == [], real.validate()
ok("real catalog validates clean with schema+meta checks")

# 2. invalid (mismatched) schema version
c = build(MAPS, {"m": [terr("m:a", "m", "a", sv=2)]}, CAT)
assert any("schemaVersion" in e and "!= catalog" in e for e in c.validate()), c.validate()
ok("territory schemaVersion mismatching catalog is rejected")

# 3. missing _meta ownership block
c = build(MAPS, {"m": [terr("m:a", "m", "a", with_meta=False)]}, CAT)
assert any("missing _meta" in e for e in c.validate()), c.validate()
ok("missing _meta ownership block is rejected")

# 4. duplicate ownership (same svg path mapped by two territories)
c = build(MAPS, {"m": [terr("m:a", "m", "a", keys=["p"]), terr("m:b", "m", "b", keys=["p"])]}, CAT)
assert any("duplicate svg path mapping" in e for e in c.validate()), c.validate()
ok("duplicate svg path ownership is rejected")

# 5. duplicate territory ids
c = build(MAPS, {"m": [terr("m:a", "m", "a"), terr("m:a", "m", "a2")]}, CAT)
assert any("duplicate territory id" in e for e in c.validate()), c.validate()
ok("duplicate territory id is rejected")

# 6. mixed schema versions across territories
c = build(MAPS, {"m": [terr("m:a", "m", "a", sv=1), terr("m:b", "m", "b", sv=2)]}, CAT)
errs = c.validate()
assert any("mixed schema versions" in e for e in errs), errs
ok("mixed schema versions are rejected")

# 7. catalog.json missing schemaVersion
c = build(MAPS, {"m": [terr("m:a", "m", "a")]}, {"catalogVersion": 1})
assert any("catalog.json missing schemaVersion" in e for e in c.validate()), c.validate()
ok("catalog.json without schemaVersion is rejected")

print("\nAll %d hardening validator tests passed." % passed)
